plugins/elantp/fu-elantp-hid-device.c

```diff
@@ -382,8 +619,48 @@ fu_elantp_hid_device_write_firmware(FuDevice *device,
 	/* write each block */
 	buf = g_bytes_get_data(fw, &bufsz);
 	iap_addr = fu_elantp_firmware_get_iap_addr(firmware_elantp);
-	chunks = fu_chunk_array_new(buf + iap_addr, bufsz - iap_addr, 0x0, 0x0, self->fw_page_size);
-	for (guint i = 0; i < chunks->len; i++) {
+
+	if (self->force_table_support &&
+	    self->force_table_addr >=
+		fu_elantp_firmware_get_forcetable_addr(FU_ELANTP_FIRMWARE(firmware))) {
+		g_autofree guint8 *buf2 = g_malloc0(bufsz);
+		if (!fu_memcpy_safe(buf2,
+				    bufsz,
+				    0x0, /* dst */
+				    buf,
+				    bufsz,
+				    0x0, /* src */
+				    bufsz,
+				    error))
+			return FALSE;
+
+		if (!fu_elantp_hid_device_filling_forcetable_firmware(
+			device,
+			buf2,
+			bufsz,
+			fu_elantp_firmware_get_forcetable_addr(FU_ELANTP_FIRMWARE(firmware)),
+			error)) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_WRITE,
+				    "filling forcetable failed");
+			return FALSE;
+		}
+		chunks = fu_chunk_array_new(buf2 + iap_addr,
+					    bufsz - iap_addr,
+					    0x0,
+					    0x0,
+					    self->fw_page_size);
+		total_pages = (self->force_table_addr - iap_addr - 1) / self->fw_page_size + 1;
```

Comment:
i'm not sure I understand this; what's going on?

Reply:
@jinglewu

Reply:
@jinglewu

Reply:
so is total_pages less than chunks->len?

Reply:
yes.

Reply:
Can you add that into the code please; it's certainly not obvious.

Reply:
Do you mean to add this 3 three comments to the code?

Reply:
I think doing something like `if (total_pages > chunks->len) g_set_error()` etc would be a good idea, but you can do this as a follow up PR if that helps.